File: hmis-engine-web/src/app/features/masterdata/wards/ward.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { environment } from '../../../../environments/environment';
import { PageResponse } from '../../../core/http/page.types';
import { CreateWardRequest, UpdateWardRequest, Ward, WardSearchParams } from './ward.types';

@Injectable({ providedIn: 'root' })
export class WardService {
  private readonly http = inject(HttpClient);
  private readonly base = `${environment.apiUrl}/masterdata/wards`;

  search(params: WardSearchParams = {}): Observable<PageResponse<Ward>> {
    let p = new HttpParams();
    if (params.query) p = p.set('query', params.query);
    if (params.active !== undefined) p = p.set('active', String(params.active));
    if (params.category) p = p.set('category', params.category);
    if (params.page !== undefined) p = p.set('page', String(params.page));
    if (params.size !== undefined) p = p.set('size', String(params.size));
    if (params.sort) p = p.set('sort', params.sort);
    return this.http.get<PageResponse<Ward>>(this.base, { params: p });
  }

  findByUid(uid: string): Observable<Ward> { return this.http.get<Ward>(`${this.base}/${uid}`); }
  create(req: CreateWardRequest): Observable<Ward> { return this.http.post<Ward>(this.base, req); }
  update(uid: string, req: UpdateWardRequest): Observable<Ward> { return this.http.put<Ward>(`${this.base}/${uid}`, req); }
  setActive(uid: string, active: boolean): Observable<Ward> { return this.http.put<Ward>(`${this.base}/${uid}/active`, { active }); }
  delete(uid: string): Observable<void> { return this.http.delete<void>(`${this.base}/${uid}`); }
}

File: hmis-engine-api/src/main/java/com/otapp/hmis/engine/masterdata/ward/application/WardService.java
package com.otapp.hmis.engine.masterdata.ward.application;

import com.otapp.hmis.engine.common.api.PageResponse;
import com.otapp.hmis.engine.common.error.ConflictException;
import com.otapp.hmis.engine.common.error.NotFoundException;
import com.otapp.hmis.engine.masterdata.ward.application.dto.CreateWardRequest;
import com.otapp.hmis.engine.masterdata.ward.application.dto.UpdateWardRequest;
import com.otapp.hmis.engine.masterdata.ward.application.dto.WardDto;
import com.otapp.hmis.engine.masterdata.ward.domain.Ward;
import com.otapp.hmis.engine.masterdata.ward.domain.WardCategory;
import com.otapp.hmis.engine.masterdata.ward.domain.WardRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class WardService {

    private final WardRepository wardRepository;

    @Transactional
    public WardDto create(CreateWardRequest request) {
        String code = request.code().trim().toUpperCase();
        if (wardRepository.existsByCode(code)) {
            throw new ConflictException("Ward code already exists: " + code);
        }
        Ward ward = new Ward(code, request.name().trim(), request.category(),
                request.capacity(), request.location(), request.description());
        wardRepository.save(ward);
        return toDto(ward);
    }

    @Transactional
    public WardDto update(String uid, UpdateWardRequest request) {
        Ward w = loadOrThrow(uid);
        w.setName(request.name().trim());
        w.setCategory(request.category());
        w.setCapacity(request.capacity());
        w.setLocation(request.location());
        w.setDescription(request.description());
        return toDto(w);
    }

    @Transactional
    public WardDto setActive(String uid, boolean active) {
        Ward w = loadOrThrow(uid);
        if (active) w.activate(); else w.deactivate();
        return toDto(w);
    }

    @Transactional
    public void delete(String uid) {
        wardRepository.delete(loadOrThrow(uid));
    }

    @Transactional(readOnly = true)
    public WardDto findByUid(String uid) {
        return toDto(loadOrThrow(uid));
    }

    @Transactional(readOnly = true)
    public PageResponse<WardDto> search(String query, Boolean active, WardCategory category, Pageable pageable) {
        String trimmed = query == null ? null : query.trim();
        return PageResponse.from(wardRepository.search(trimmed, active, category, pageable).map(WardService::toDto));
    }

    private Ward loadOrThrow(String uid) {
        return wardRepository.findByUid(uid).orElseThrow(() -> new NotFoundException("Ward not found: " + uid));
    }

    private static WardDto toDto(Ward w) {
        return new WardDto(w.getUid(), w.getCode(), w.getName(), w.getCategory(),
                w.getCapacity(), w.getLocation(), w.getDescription(), w.isActive(),
                w.getCreatedAt(), w.getUpdatedAt());
    }
}
